### PR TITLE
Add support for more archive formats

### PR DESCRIPTION
**Pull request by SeekingMeaning**
Not sure if this works, it seems that gofish is still using the old unarchiving code

My setup:

```
cd %USERPROFILE%\go\src\github.com\fishworks
git clone https://github.com/fishworks/gofish
cd gofish
dep ensure
cd cmd\gofish
go run . install cabal-install
```

### CODE REVIEW

File: food.go [hunk 1]
```diff
@@ -182,11 +181,9 @@ func unarchiveOrCopy(src, dest, urlPath string) error {
 	}
 	defer in.Close()
 
-	if archive.IsArchivePath(src) {
-		return archive.Untar(in, dest, &archive.TarOptions{NoLchown: true})
-	} else if isZipPath(src) {
-		in.Close()
-		return unzip(src, dest)
+	err = archiver.Unarchive(src, dest)
```
Comment (bacongobbler) [inline food.go]: this line and the following line can be simplified to

```
if err := archiver.Unarchive(src, dest); err != nil {
```

Would you mind making that change?

**Comment by bacongobbler** [comment]
can you remove lines 178-182? it looks like that is no longer necessary.

**Comment by bacongobbler** [comment]
looks great! This also adds support for bzip2 and rar formats. This is great! Thanks for looking into this. Just some small revisions then looks good to go.

**Comment by SeekingMeaning** [comment]
Oh, it looks like `in` is still needed

**Comment by bacongobbler** [comment]
hmm, right. Instead, can you move lines 178-182 after the `if` block? That code is necessary to copy the content if it is not in an archived format. Previously it was needed for the `zip` format to unarchive the file, but this PR removed that bit of code so it can be shifted after the archive check.

**Comment by bacongobbler** [comment]
...which you have just done. 😄  Thanks!

File: food.go [hunk 2]
```diff
@@ -176,18 +175,15 @@ func (f *Food) Uninstall() error {
 }
 
 func unarchiveOrCopy(src, dest, urlPath string) error {
+	if err := archiver.Unarchive(src, dest); err == nil {
```
Comment (bacongobbler) [inline food.go]: how do we handle the case where the unarchiving process failed, like if the archive was corrupted, or the file extension did not match the archive format?

Comment (SeekingMeaning) [reply food.go]: Hm... We might be able to use the `ByExtension` function, like here: https://github.com/mholt/archiver/blob/2d14498067935bc817c9100490782583fb5ce44e/archiver.go#L198-L210

Comment (SeekingMeaning) [reply food.go]: Maybe something like this:

```go
if uaIface, err := archiver.ByExtension(source); err == nil {
	u, ok := uaIface.(Unarchiver)
	if !ok {
		return fmt.Errorf("format specified by source filename is not an archive format: %s (%T)", source, uaIface)
	}
	return u.Unarchive(source, destination)
}
```

or this:

```go
if uaIface, err := archiver.ByExtension(source); err == nil {
	if u, ok := uaIface.(Unarchiver); ok {
		return u.Unarchive(source, destination)
	}
}
```

**Pull request closed without merging**